### PR TITLE
fix(ehttpc): fix retry

### PR DESCRIPTION
**Pull request by tigercl**

### CODE REVIEW

**Comment by zmstone** [comment]
please add more description to the PR and the commit message.
what issue is this solving

File: src/ehttpc.erl [hunk 1]
```diff
@@ -58,35 +58,30 @@
...
+
+request({Pool, N}, Method, Request, Timeout, Retry) when is_atom(Pool) ->
+    request(ehttpc_pool:pick_worker(Pool, N), Method, Request, Timeout, Retry);
+
 request(Worker, Method, Request, Timeout, Retry) when is_pid(Worker) ->
     ExpireAt = now_() + Timeout,
     try gen_server:call(Worker, {Method, Request, ExpireAt}, Timeout + 1000) of
         %% gun will reply {gun_down, _Client, _, normal, _KilledStreams, _} message
         %% when connection closed by keepalive
         {error, Reason} ->
```
Comment (zmstone) [inline src/ehttpc.erl:78]: `when NRetry =< 1 ->`

File: src/ehttpc.erl [hunk 2]
```diff
@@ -58,35 +58,30 @@
...
     request(ehttpc_pool:pick_worker(Pool), Method, Request, Timeout, Retry);
-request({Pool, KeyOrNum}, Method, Request, Timeout, Retry) when is_atom(Pool) ->
-    request(ehttpc_pool:pick_worker(Pool, KeyOrNum), Method, Request, Timeout, Retry);
-request(Worker, _Method, _Req, _Timeout, 0) when is_pid(Worker) ->
-    {error, normal};
+
+request({Pool, N}, Method, Request, Timeout, Retry) when is_atom(Pool) ->
+    request(ehttpc_pool:pick_worker(Pool, N), Method, Request, Timeout, Retry);
+
 request(Worker, Method, Request, Timeout, Retry) when is_pid(Worker) ->
```
Comment (zmstone) [inline src/ehttpc.erl:73]: use a different function name if all calls need to hit line 70 first.

**Comment by tigercl** [comment]
Mainly to fix the problem that the retry mechanism did not return the cause of the error as expected, so the maximum number of retries was reached, and the actual cause of the error was instead covered by normal.

The other is mainly the adjustment of the code style. In fact, I only want the request interface to accept parameter `Pool`, but in order to ensure the hot upgrade, it must not continue to be compatible with `Worker`, so you will see the original parameter named `PoolOrWorker`.

However, I did encounter a problem. When the pool type is random, I can only pass in the parameter `Pool`, but when the pool type is hash, I must also pass in parameter `N`. I didn’t expect it to be suitable. Method. Maybe just receiving `Worker` parameters is the correct approach?

**Comment by zmstone** [comment]
maybe treat warning as error. because:
```
src/ehttpc.erl:64: Warning: this clause for request/3 cannot match because a previous clause at line 61 always matches
src/ehttpc.erl:70: Warning: this clause for request/4 cannot match because a previous clause at line 67 always matches
```

**Review by zmstone — APPROVED** [review]

**Comment by tigercl** [comment]
> maybe treat warning as error. because:
> 
> ```
> src/ehttpc.erl:64: Warning: this clause for request/3 cannot match because a previous clause at line 61 always matches
> src/ehttpc.erl:70: Warning: this clause for request/4 cannot match because a previous clause at line 67 always matches
> ```

I did not find this error when compiling locally.

**Comment by zmstone** [comment]
> > maybe treat warning as error. because:
> > ```
> > src/ehttpc.erl:64: Warning: this clause for request/3 cannot match because a previous clause at line 61 always matches
> > src/ehttpc.erl:70: Warning: this clause for request/4 cannot match because a previous clause at line 67 always matches
> > ```
> 
> I did not find this error when compiling locally.

There were such warnings before your change.
And my commit added warnings as error to rebar config

**Pull request merged**